### ISSUE TITLE
Extract core logic to allow different front-ends or platforms

### ISSUE THREAD

**Issue by HoLLy-HaCKeR** [inorichi/tachiyomi#2119]
Extract the core logic and extension contracts into a different project, allowing for Tachiyomi to be ported to different platforms such as Desktop or Web.

Structure could be like this (using a C# naming scheme):
```yaml
core:             # provide core logic, extension interfaces
    - Tachiyomi.Lib
platform support: # provide abstractions over platform-specific APIs, user interfaces
    - Tachiyomi.Platform.Android
    - Tachiyomi.Platform.Desktop
plugins:
    - ...
```

Doing this also allows third-party developers to use the Tachiyomi engine into their applications, providing different user interfaces (eg. under a web server as proposed in #2118) or support for other platforms entirely.

I'm also not very familiar with the existing project structure and how Kotlin/Java handles multiple projects (I'm from a C# background), so I don't know exactly how much effort with would go into this.

**Comment by arkon** [comment]
Again, see https://github.com/TachiWeb . However, its implementation is dependant on how 1.x of Tachiyomi is designed.

**Comment by frnco** [comment]
+1 here.

I actually started doing something quite similar to extracting, mostly I'm creating an API and Tachiyome is one of my biggest references.

Since the ideas kind of overlap, it might be nice to collaborate. I started doing it in JS, mostly because I can call node modules from pretty much anywhere, but it's probably better to do it in Kotlin if there's a chance you guys are also interested.

Also, one of my main points, I wanna make it VERY easy to add/remove sources, and make it so people can edit the sources, so that we can fix stuff ourselves when it breaks, send PRs or whatever. I've had some breakages and I know it really sucks.
